lib/storage/environment.go

```diff
@@ -107,9 +109,60 @@ func (r *EnvironmentV1) GetKeyValues() map[string]string {
 
 // CheckAndSetDefaults validates this resource and sets defaults
 func (r *EnvironmentV1) CheckAndSetDefaults() error {
+	var errors []error
+	if err := r.checkHTTPProxy(); err != nil {
+		errors = append(errors, err)
+	}
+	if err := r.checkHTTPSProxy(); err != nil {
+		errors = append(errors, err)
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// checkHTTPProxy verifies HTTP_PROXY (if present) is valid a URL and has http:// scheme.
+func (r *EnvironmentV1) checkHTTPProxy() error {
```

Comment:
These validation rules don't appear to match atleast golang validation logic: https://github.com/golang/go/blob/master/src/vendor/golang.org/x/net/http/httpproxy/proxy.go#L149-L170

1. it appears the scheme isn't required, and defaults to http.
2. socks5 isn't considered a valid scheme.
3. it's valid to use an http proxy for https traffic. And an https proxy for http traffic.

Also I'm actually confused about the originally stated problem now, since it looks like the golang validation logic just drops any parsing errors. I must be missing something.
https://github.com/golang/go/blob/master/src/vendor/golang.org/x/net/http/httpproxy/proxy.go#L211-L216

Reply:
@knisbet I thought that the original problem was that some software didn't like the fact that proxy URL is missing the scheme?

Reply:
@knisbet I've relaxed the requirements to only check for the presence of a valid scheme (http/https/socks5) in the provided proxy address to behave more like the Go parse function. Unlike the Go parser though, I have decided to make the scheme mandatory rather than assuming "http" - I think it may cause more trouble than help, as after some experimentation is looks like that `url.Parse()` will parse almost anything without an error, so if we try to prepend "http" everywhere, it will happily let some bogus address through as well. So, I think that having a bit stricter requirements to have scheme explicitly will help us filter out potentially bad addresses better.

Reply:
> @knisbet I thought that the original problem was that some software didn't like the fact that proxy URL is missing the scheme?

I thought so as well, just I don't see how now. I suppose unless it's failing somewhere later down the line, not in parsing the proxy, but when it actually tries to use it, throws an unexpected error. In which case would a valid proxy that isn't working possibly cause problems for docker?

> Unlike the Go parser though, I have decided to make the scheme mandatory rather than assuming "http" - I think it may cause more trouble than help, as after some experimentation is looks like that url.Parse() will parse almost anything without an error, so if we try to prepend "http" everywhere, it will happily let some bogus address through as well.

I dunno. I think our objective on this one is not to crash on invalid configuration. But introducing requirements beyond what is normally accepted does create some of it's own risks. Does mandating the scheme really make the url.Parse method more strict, or will it just require a bogus address to contain a scheme?

Reply:
Yeah it's a bit tough without actually seeing the logs of the alleged Docker crash due to the missing scheme.

You're right that mandating the scheme doesn't really make url.Parse() more strict (if someone wants to break something, they will), but I do think that it reduces the chance of accidental error in user's proxy configuration.

Just looking online, pretty much every single example of http_proxy includes an explicit scheme - the fact that Go/curl default to http seems more of an "assumption" (and not a great one if you ask me) and requiring the scheme would also protect us in case some software does actually require it.